lib/lolcat/lol.rb

```diff
@@ -47,7 +46,8 @@ def self.cat(fd, opts={})
       begin
         begin
           buf += fd.sysread(4096)
-        end while buf.match(INCOMPLETE_ESCAPE) or (fd.external_encoding == 'UTF-8' and buf.match(INCOMPLETE_UTF8))
+          invalid_encoding = !buf.dup.force_encoding(fd.external_encoding).valid_encoding?
```

Comment:
`String#dup` uses shared strings and we avoid a regex scan so this should improve performance.